### ISSUE TITLE
feat(JSDoc): Recognize @type

### ISSUE THREAD

**Issue by Conaclos** [eslint/eslint#7325]
# Problem

JSDoc enabled to use [`@type`](http://usejsdoc.org/tags-type.html) to specify the type of a property/variable, including callbacks:

``` js
/**
 * @type {function(a: Object)}
 */
let onEvent = (a) => {}
```

Currently, ESLint reports the last code as invalid:

```
error | Missing JSDoc for parameter 'a' | valid-jsdoc
```
# Proposition

If the documentation includes a tag `@type` then the use of the tag `@param` is optional.

If at least one tag `@param` is available then check the JSDoc like now (every parameter must have its `@param` tag, ...).
# Environment

ESLint 3.7.x

EDIT: I'm using a configuration available [here](https://raw.githubusercontent.com/Conaclos/eslint-config-conaclos/master/base.js). The option "requireReturn" is disabled.

**Comment by kaicataldo** [comment]
I'm not seeing any examples for this specific syntax in the JSDoc docs - are you sure this is correct? Specifically, I'm having difficulty finding examples of the Flow-like parameter syntax in your example: `(a: Object)`.

Also, I'm getting two errors for the same example:
```
 1:1  error  Missing JSDoc @returns for function  valid-jsdoc
 1:1  error  Missing JSDoc for parameter 'a'      valid-jsdoc
```

**Comment by Conaclos** [comment]
> Also, I'm getting two errors for the same example

Yes, I've just edited the issue. I forgot to note that I disabled a default option...

> I'm not seeing any examples for this specific syntax in the JSDoc docs - are you sure this is correct? I'm having difficulty finding examples of the Flow-like parameter syntax

You are right, I'm using an [ESDoc](https://esdoc.org/) syntax.
I think so, in JSDOc, this is more likely to be:

```
/**
 * @callback onEventCb
 * @param {Object} a
 */

/**
 * @type {onEventCb}
 */
let onEvent = (a) => {}
```

However it is not changing the core of the proposition :)

**Comment by kaicataldo** [comment]
Thanks for clarifying - it's difficult for us to discuss if we don't have a clear understanding of what the proposal is :)

**Comment by kaicataldo** [comment]
Closing this as discussion has stalled and it doesn't appear there is enough interest, either from the team or from the community, to continue the discussion at this time. While we wish we could accommodate all requests, we have limited resources and do need to prioritize. Please feel free to open a new issue with a proposal to continue this discussion if this is something you would like to continue advocating for.
